### PR TITLE
feat: ws implementation

### PR DESCRIPTION
**Pull request by lewxdev**

### CODE REVIEW

File: app/page.tsx [hunk 1]
```diff
@@ -2,78 +2,70 @@
...
+      )}
+      onClick={() => {
+        socket.emit("expose", index);
+      }}
+      onContextMenu={(event) => {
+        event.preventDefault();
+        socket.emit("flag", index);
+      }}
+    >
+      {state !== "flagged" && state !== "unknown" && state !== "mine" && state}
```
Comment (drewradcliff) [inline app/page.tsx]: typeof
```suggestion
      {typeof state === "number" && state}
```

File: app/page.tsx [hunk 2]
```diff
@@ -2,78 +2,70 @@
...
-    function onMessage(message: string) {
-      setMessages((messages) => [...messages, message]);
-    }
-
-    socket.on("message", onMessage);
-
-    return () => {
-      socket.off("message", onMessage);
-    };
+    socket.on("update", setCells);
```
Comment (drewradcliff) [inline app/page.tsx:51]: add cleanup function

File: app/page.tsx [hunk 3]
```diff
@@ -2,78 +2,70 @@
...
+          gridTemplateRows: `repeat(${size}, ${GRID_SIZE}em)`,
         }}
       >
-        {field["data"].map((n, index) => {
-          const value = n % 10;
-          const color = colorMap.get(value)!;
-          return <div className={color} key={index} />;
-        })}
+        {cells?.map((state, index) => (
+          <Cell index={index} key={index} state={state} />
```
Comment (drewradcliff) [inline app/page.tsx:70]: ideally don't use index as key here

**Review by drewradcliff — APPROVED** [review]
Nice work 💪

File: app/page.tsx (hunk 3 above)
Comment (lewxdev) [reply app/page.tsx:70]: "I think it's fine for now"

**Pull request merged**